File: wun-android/src/main/kotlin/wun/foundation/BadgeRenderer.kt
// Compose renderer for `:wun/Badge`. `:tone` picks a semantic colour
// (info / success / warning / danger). Unknown tones fall back to
// info so server-side additions don't break older clients.

package wun.foundation

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.serialization.json.JsonPrimitive
import wun.WunComponent

object BadgeRenderer {
    val render: WunComponent = { props, children ->
        val tone = (props["tone"] as? JsonPrimitive)
            ?.takeIf { it.isString }?.content ?: "info"
        val (bg, fg) = when (tone) {
            "success" -> Color(0xFFE1F6E9) to Color(0xFF1B6C3A)
            "warning" -> Color(0xFFFFF3D6) to Color(0xFF7A5300)
            "danger"  -> Color(0xFFFDE2E2) to Color(0xFF9B1C1C)
            else      -> Color(0xFFE8F1FF) to Color(0xFF0A4EA3)
        }
        val label = Children.flatten(children)
        Text(
            text       = label,
            color      = fg,
            fontSize   = 11.sp,
            fontWeight = FontWeight.SemiBold,
            modifier   = Modifier
                .clip(RoundedCornerShape(999.dp))
                .background(bg)
                .padding(horizontal = 8.dp, vertical = 2.dp),
        )
    }
}

File: wun-ios/Sources/Wun/Foundation/BadgeRenderer.swift
// SwiftUI renderer for `:wun/Badge`. `:tone` picks a semantic colour
// (info / success / warning / danger). Falls back to `info` for any
// unknown tone string so the server can introduce new tones without
// hard-breaking older clients.

import SwiftUI

public enum WunBadge {
    public static let render: WunComponent = { props, children in
        let tone = props["tone"]?.stringValue ?? "info"
        let label = WunChildren.flatten(children)
        let (bg, fg): (Color, Color) = {
            switch tone {
            case "success": return (Color(red: 0.88, green: 0.96, blue: 0.91),
                                    Color(red: 0.10, green: 0.42, blue: 0.23))
            case "warning": return (Color(red: 1.00, green: 0.95, blue: 0.84),
                                    Color(red: 0.48, green: 0.32, blue: 0.00))
            case "danger":  return (Color(red: 0.99, green: 0.89, blue: 0.89),
                                    Color(red: 0.61, green: 0.11, blue: 0.11))
            default:        return (Color(red: 0.91, green: 0.94, blue: 1.00),
                                    Color(red: 0.04, green: 0.31, blue: 0.64))
            }
        }()
        return AnyView(
            SwiftUI.Text(label)
                .font(.system(size: 11, weight: .semibold))
                .padding(.horizontal, 8)
                .padding(.vertical, 2)
                .background(bg)
                .foregroundColor(fg)
                .cornerRadius(999)
        )
    }
}
